### ISSUE TITLE
Link generation does not work in <pre> blocks

### ISSUE THREAD

**Issue by palemieux** [w3c/respec#1970]
## Important info

```
<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <script src='https://www.w3.org/Tools/respec/respec-w3c-common' async class='remove'></script>
    <script class='remove'>
      var respecConfig = {
        specStatus: "unofficial"
      };
    </script>
  </head>
  <body>
    <h1>My Awesome API</h1>
    <section id='abstract'>
      <p>
        This is required.
      </p>
    </section>
    <section id='sotd'>
      <p>
        This is required.
        <pre>
        <a>Foo</a>
        <a href="w3.org">W3</a></pre>
      </p>
      <p><dfn>Foo</dfn> is great.</p>
      <p>Did I mention <a>Foo</a>.</p>
    </section>
  </body>
</html>
```

## Description of problem

```<a>Foo</a>``` in `<pre>` is not changed into a link to the definition of `Foo`.

**Comment by marcoscaceres** [comment]
To link, pre would need to have class=idl (and would need to be valid WebIDL).

**Comment by marcoscaceres** [comment]
Sorry, I misunderstood. It’s removing the link? I’ll take a look.

**Comment by saschanaz** [comment]
Looks like there is a regression in core/structure.

**Comment by devanshbatra04** [comment]
I investigated it, the issue seems to be in ```core/highlight``` plugin. It run perfectly well without including it. I will see more into the plugin to find exact cause.
 
#### After excluding ```core/highlight```
![screenshot from 2018-12-22 20-33-00](https://user-images.githubusercontent.com/33961252/50375773-ed3a7300-0628-11e9-90ab-cc82af8eed9f.png)

**Comment by marcoscaceres** [comment]
@devanshbatra04, do you want to take this one?

**Comment by devanshbatra04** [comment]
Yes definitely!

**Comment by pradeepgangwar** [comment]
@devanshbatra04 Any update? If your are not working then I would like to take this up.

**Comment by devanshbatra04** [comment]
Not been well at all. I will work on it and the other issues asap. Please have a look other issues. :)

**Comment by CodHeK** [comment]
@devanshbatra04 mind if I start working on this issue ?

**Comment by marcoscaceres** [comment]
@CodHeK feel free to take it.

**Comment by CodHeK** [comment]
@marcoscaceres this is where the highlight is added right ? [https://github.com/w3c/respec/blob/develop/src/core/highlight.js#L62](https://github.com/w3c/respec/blob/develop/src/core/highlight.js#L62)

**Comment by CodHeK** [comment]
Also, what I noticed is that when it's selecting all the elements to be linked, [here](https://github.com/w3c/respec/blob/develop/src/core/link-to-dfn.js#L35) it is not able to select `<a>` elements that are inside the `<pre>` tag!

**Comment by devanshbatra04** [comment]
@CodHeK ```core/link-to-dfn.js``` is alright, it does include this links and is not the source of this issue.
You have identified correctly that the highlights are added at https://github.com/w3c/respec/blob/develop/src/core/highlight.js#L62 
However on the line preceding this (on #L61) The ```innerHTML``` gets rewritten for the ```<pre>``` element from the message received from https://github.com/w3c/respec/blob/develop/src/core/highlight.js#L48

If you look closely, ```code: element.textContent``` does not include the html anchor tags, thus only the textContent is displayed (and not the HTML) 
That's where you need to make some small change, I tried using innerHTML isntead of textConent, but that doesn't work because it gets parsed later to text in the ```value``` field somehow.

**Comment by marcoscaceres** [comment]
And then here:
https://github.com/w3c/respec/blob/develop/worker/respec-worker.js#L22

**Comment by devanshbatra04** [comment]
Wow I had been looking at the wrong worker all this while :"(

**Comment by CodHeK** [comment]
Also, the problem here is that it's taking `Foo W3C` as a single text, probably why it's not able to highlight `Foo` to its `dfn` ..

**Comment by CodHeK** [comment]
> If you look closely, `code: element.textContent` does not include the html anchor tags, thus only the textContent is displayed (and not the HTML)

Yeah, this is the problem! Probably, we'd have to have to save `element.childNodes` and insert them later

**Comment by CodHeK** [comment]
@devanshbatra04 commenting out `element.innerHTML` fixed the bug!

**Comment by devanshbatra04** [comment]
that's not a fix. that's undoing anything ```core/highlight``` was doing in the first place.

**Comment by CodHeK** [comment]
Why do we need to assign innerHTML = value in the first place ?


On Sunday, February 17, 2019, 3:20 PM, Devansh Batra <notifications@github.com> wrote:


that's not a fix. that's undoing anything core/highlight was doing in the first place.

**Comment by saschanaz** [comment]
Maybe we should run the highlighter first and then do linking?

**Comment by devanshbatra04** [comment]
@saschanaz wouldn't work because the highlighter removes anchor tags, just keeps the ```innerText``` of the ```<pre>``` block. 
I think we need to rewrite ```respec-worker```

**Comment by saschanaz** [comment]
@devanshbatra04 I tried the example in the OP and found the highlighter keep the anchor. https://codepen.io/SaschaNaz/pen/daQBLG

**Comment by saschanaz** [comment]
@devanshbatra04 I retried locally and okay, you're right.

**Comment by saschanaz** [comment]
So I would suggest two ways to fix (or workaround) this:

1. Just use `class=nohighlight`
2. Or, to make things strict, require a valid language hint (`class="(language name)"`) to run highlighter.

**Comment by saschanaz** [comment]
>I think we need to rewrite `respec-worker` to use `innerHTML` instead

https://github.com/w3c/respec/blob/4e9f304f522fe03c21e36c52f732a2abf2908c5c/src/core/highlight.js#L48-L53

That means here. Yeah, maybe we could instead pass `element.innerHTML` and the highligher may keep the anchor intact... I don't have a strong opinion here.

**Comment by devanshbatra04** [comment]
![screenshot from 2019-02-17 16-10-17](https://user-images.githubusercontent.com/33961252/52911773-8b1b8880-32ce-11e9-8af0-77fa27da2fd4.png)
Here's what happens if I just pass the ```element.innerHTML```, as I said we need to rewrite respec-worker using this approach. I do like ```class=nohighlight``` instead.

**Comment by devanshbatra04** [comment]
```html
<pre class="nohighlight">
        <a>Foo</a>
        <a href="w3.org">W3</a>
</pre>
```
This code is working fine already, so I believe the issue stands resolved?

**Comment by devanshbatra04** [comment]
Although I believe we should still use ```element.innerHTML``` to allow support for HTML.

**Comment by marcoscaceres** [comment]
I think I'm also leaning towards what @saschanaz said in https://github.com/w3c/respec/issues/1970#issuecomment-464439658 (i.e., the author should use  `class="nohighlight"`) as otherwise this mixes programming languages with spec stuff. 

@palemieux can you live with the above? I.e., we can't really fix this because ReSpec depends on an external library to do the syntax highlighting.

**Comment by marcoscaceres** [comment]
closing as "won't fix".

**Comment by saschanaz** [comment]
@marcoscaceres 

[The HTML spec](https://html.spec.whatwg.org/multipage/text-level-semantics.html#the-code-element) recommends using `<pre>` and `<code>` together. Should we follow it? (Related issues: https://github.com/whatwg/html/issues/3764, https://github.com/whatwg/html/pull/3768)

If then we could fix this by requiring `<code>` inside `<pre>` to run highlighter, although it would be a breaking change...

**Comment by marcoscaceres** [comment]
I'd prefer not to do that... but we should make our output be HTML-spec conforming (ie., `<pre><code>`, which is currently isn't). 

So would prefer authors just keep using `<pre class="lang">` everywhere.

**Comment by marcoscaceres** [comment]
files https://github.com/w3c/respec/issues/2098

**Comment by saschanaz** [comment]
>So would prefer authors just keep using `<pre class="lang">` everywhere.

I'm okay with that but the issue here is about `<pre>` without a language tag. Maybe we could:

1. If `<pre>` has a class, keep the current behavior
2. If it has no class and no `<code>` in it, do not run highlighter for the element.

**Comment by marcoscaceres** [comment]
> 2\. If it has no class and no `<code>` in it, do not run highlighter for the element.

We might need to check some specs to see how much breakage there would be. The assumption thus far with the highlighter is that it automatically detect the language for you (even if it's a bit hit or miss).

**Comment by CodHeK** [comment]
Does this mean, if the `<pre>` element has `class="nohighlight"` we need not wrap the contents of the `pre` inside `code` ? And wrap it other case when some other class or no class is used for the pre tag.

**Comment by saschanaz** [comment]
This issue is closed so I'll comment in #2098.

**Comment by CodHeK** [comment]
I'll just paste the comment in #2098
